tests/e2e/tidbcluster/serial.go

```diff
@@ -589,6 +589,229 @@ var _ = ginkgo.Describe("[Serial]", func() {
+			tc.Spec.TiKV.Replicas = 1
+			tc.Spec.TiDB.Replicas = 1
+
+			err := genericCli.Create(context.TODO(), tc)
+			framework.ExpectNoError(err, "Expected TiDB cluster created")
+			err = oa.WaitForTidbClusterReady(tc, 6*time.Minute, 5*time.Second)
+			framework.ExpectNoError(err, "Expected TiDB cluster ready")
+
+			getPods := func(ls string) ([]v1.Pod, error) {
+				listOptions := metav1.ListOptions{
+					LabelSelector: ls,
+				}
+				podList, err := c.CoreV1().Pods(ns).List(listOptions)
+				if err != nil {
+					return nil, err
+				}
+				return podList.Items, nil
+			}
+
+			tidbPods, err := getPods(labels.SelectorFromSet(label.New().Instance(tcName).TiDB().Labels()).String())
+			framework.ExpectNoError(err, "failed to get tidb pods")
+
+			ginkgo.By("Set --selector=version=old for the default TiDB Operator")
+			ocfg.Selector = []string{"version=old"}
+			ocfg.Tag = cfg.OperatorTag
+			ocfg.Image = cfg.OperatorImage
+			oa.InstallCRDOrDie(ocfg)
+			oa.UpgradeOperatorOrDie(ocfg)
+			log.Logf("Upgrade operator with --set-string \"selector=version=old\"")
+
+			ginkgo.By("Upgrade TiDB of TidbCluster 1, wait for 1min, check that no rolling update occurs")
+			err = controller.GuaranteedUpdate(genericCli, tc, func() error {
+				tc.Spec.Version = utilimage.TiDBV4
+				return nil
+			})
+			framework.ExpectNoError(err, "failed to update TidbCluster 1 to upgrade tidb version to %v", utilimage.TiDBV4)
+
+			err = wait.Poll(5*time.Second, 1*time.Minute, func() (done bool, err error) {
+				// confirm the tidb haven't been changed
+				changed, err := utilpod.PodsAreChanged(c, tidbPods)()
+				if err != nil {
+					log.Logf("ERROR: meet error during verify tidb pods, err:%v", err)
+					return true, nil
+				}
+				if changed {
+					return true, nil
+				}
+				log.Logf("confirm tidb pods haven't been changed this time")
+
+				return false, nil
+			})
+			framework.ExpectEqual(err, wait.ErrWaitTimeout, "expect pd/tikv/tidb haven't been changed for 5 minutes")
+			log.Logf("Upgrade TidbCluster 1 with new version, but not rolling update.")
+
+			ginkgo.By("Set label version=old to TidbCluster 1, check that TiKV of TidbCluster 1 has been rolling updated")
+			err = controller.GuaranteedUpdate(genericCli, tc, func() error {
+				tc.Labels = map[string]string{"version": "old"}
+				return nil
+			})
+			framework.ExpectNoError(err, "failed to update TidbCluster 1 to set lable version=old")
+
+			err = wait.Poll(5*time.Second, 1*time.Minute, func() (done bool, err error) {
+				// confirm the tidb haven't been changed
+				changed, err := utilpod.PodsAreChanged(c, tidbPods)()
+				if err != nil {
+					log.Logf("ERROR: meet error during verify tidb pods, err:%v", err)
+					return true, nil
+				}
+				if changed {
+					return true, nil
+				}
+				log.Logf("confirm tidb pods haven't been changed this time")
+
+				return false, nil
+			})
+			framework.ExpectEqual(err, wait.ErrWaitTimeout, "expect pd/tikv/tidb have been changed for 5 minutes")
```

Comment:
Label has been added to the TC, so the Pods should be rolling updated, why expect timeout error here?

Reply:
Maybe refer to the logic here https://github.com/pingcap/tidb-operator/pull/3844/files#diff-c2f1c23cb63d938919638cead8a132ce1ef660fb1814f2db6c045f34784e747dR1711 to check the Pods have been upgraded.